lib/actions/local/flash.ts

```diff
@@ -17,17 +17,17 @@ limitations under the License.
 import { CommandDefinition } from 'capitano';
 import chalk from 'chalk';
 import { stripIndent } from 'common-tags';
-import * as sdk from 'etcher-sdk';
+import * as SDK from 'etcher-sdk';
 
 async function getDrive(options: {
 	drive?: string;
-}): Promise<sdk.sourceDestination.BlockDevice> {
+}): Promise<SDK.sourceDestination.BlockDevice> {
 	const sdk = await import('etcher-sdk');
 
 	const adapter = new sdk.scanner.adapters.BlockDeviceAdapter(() => false);
 	const scanner = new sdk.scanner.Scanner([adapter]);
 	await scanner.start();
-	let drive: sdk.sourceDestination.BlockDevice;
+	let drive: SDK.sourceDestination.BlockDevice;
```

Comment:
Interesting to note that even inside the functions that declare the lowercase 'sdk' constant, I had to use the global uppercase 'SDK' in type expressions like this, to avoid compilation errors. The goal of avoiding a global sdk import in compiled js code was achieved.

Reply:
Yes, that's because the compiler checks to see whether the top level imported module (`import * as SDK from 'etcher-sdk';`) is ever used as part of an expression, and if not it uses it purely as a type identifier (ie. an underlying `require` is never called and therefore it doesn't need to emit this in built code). The type of `sdk` in your `await import('etcher-sdk')` is actually of type `SDK`. As this is the type namespace for `sdk`, you need to use the `SDK` type for any exported types/values. Hence `drive` must be of `SDK.sourceDestination.BlockDevice` and not `sdk.sourceDestination.BlockDevice`.

Reply:
I'd be surprised if it didn't do multi-pass typing, I think most modern compilers do. The only other thing I can think of is that because it's a class type rather then a function (in TypeScript), that it can retrieve the typings from an instantiation member. Or maybe there's a bug. :)

On the topic of `typeof`, it's worth reading the 'Optional Module Loading and Other Advanced Loading Scenarios' here: https://www.typescriptlang.org/docs/handbook/modules.html I believe the difference is that TS is a bit more savvy now, and can imply `instance: typeof <type>` from `instance: <type>`. I'm pretty certain I hit this a couple of years ago, and it couldn't (though entirely possible I'm misremembering).

Reply:
Maybe upgrading to TS 2.9 can help improving this, since it added *import types*
See: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html

We should definitely plan for a TS upgrade anyway, since the [v12 balena-sdk](https://github.com/balena-io/balena-sdk/tree/v12-dev) will be using TS 3.0.3, in order to be able to use some of the newer features, to better type the pine options.